File: android/src/main/kotlin/com/native_binder/StandardMessageCodec.kt
package com.native_binder

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.charset.StandardCharsets
import java.util.ArrayList
import java.util.HashMap

/**
 * Kotlin implementation of Flutter's StandardMessageCodec binary format
 * (subset: null, bool, int, long, double, String, List, Map).
 */
internal object StandardMessageCodec {
    private const val NULL: Byte = 0
    private const val TRUE: Byte = 1
    private const val FALSE: Byte = 2
    private const val INT: Byte = 3
    private const val LONG: Byte = 4
    private const val DOUBLE: Byte = 6
    private const val STRING: Byte = 7
    private const val LIST: Byte = 12
    private const val MAP: Byte = 13

    /** Reads a single value from [buffer] (used for multi-value messages). */
    fun readValueFromBuffer(buffer: ByteBuffer): Any? = readValue(buffer)

    fun encodeMessage(value: Any?): ByteArray {
        val stream = ByteArrayOutputStream()
        writeValue(stream, value)
        return stream.toByteArray()
    }

    /** Writes a single value to an existing stream (for multi-value messages). */
    fun writeValueToStream(stream: ByteArrayOutputStream, value: Any?) {
        writeValue(stream, value)
    }

    private fun writeSize(stream: ByteArrayOutputStream, value: Int) {
        when {
            value < 254 -> stream.write(value)
            value <= 0xFFFF -> {
                stream.write(254)
                stream.writeShort(value)
            }
            else -> {
                stream.write(255)
                stream.writeInt(value)
            }
        }
    }

    private fun writeValue(stream: ByteArrayOutputStream, value: Any?) {
        when (value) {
            null -> stream.write(NULL.toInt())
            is Boolean -> stream.write(if (value) TRUE.toInt() else FALSE.toInt())
            is Int -> {
                stream.write(INT.toInt())
                stream.writeInt(value)
            }
            is Long -> {
                stream.write(LONG.toInt())
                stream.writeLong(value)
            }
            is Double, is Float -> {
                stream.write(DOUBLE.toInt())
                stream.writeAlignment(8)
                stream.writeLong(java.lang.Double.doubleToLongBits((value as Number).toDouble()))
            }
            is Number -> {
                val v = value.toLong()
                if (v >= Int.MIN_VALUE && v <= Int.MAX_VALUE) {
                    stream.write(INT.toInt())
                    stream.writeInt(v.toInt())
                } else {
                    stream.write(LONG.toInt())
                    stream.writeLong(v)
                }
            }
            is String -> {
                stream.write(STRING.toInt())
                val utf8 = value.toByteArray(StandardCharsets.UTF_8)
                writeSize(stream, utf8.size)
                stream.write(utf8)
            }
            is List<*> -> {
                stream.write(LIST.toInt())
                writeSize(stream, value.size)
                value.forEach { writeValue(stream, it) }
            }
            is Map<*, *> -> {
                stream.write(MAP.toInt())
                writeSize(stream, value.size)
                value.forEach { (k, v) ->
                    writeValue(stream, k)
                    writeValue(stream, v)
                }
            }
            else -> throw IllegalArgumentException("Unsupported type: ${value::class.java}")
        }
    }

    private fun readSize(buffer: ByteBuffer): Int {
        if (!buffer.hasRemaining()) throw IllegalArgumentException("Message corrupted")
        val b = buffer.get().toInt() and 0xFF
        return when (b) {
            254 -> buffer.short.toInt() and 0xFFFF
            255 -> buffer.int
            else -> b
        }
    }

    private fun readValue(buffer: ByteBuffer): Any? {
        if (!buffer.hasRemaining()) throw IllegalArgumentException("Message corrupted")
        val type = buffer.get()
        return readValueOfType(type, buffer)
    }

    private fun readValueOfType(type: Byte, buffer: ByteBuffer): Any? {
        return when (type) {
            NULL -> null
            TRUE -> true
            FALSE -> false
            INT -> buffer.int
            LONG -> buffer.long
            DOUBLE -> {
                readAlignment(buffer, 8)
                buffer.double
            }
            STRING -> {
                val len = readSize(buffer)
                val bytes = ByteArray(len)
                buffer.get(bytes)
                String(bytes, StandardCharsets.UTF_8)
            }
            LIST -> {
                val size = readSize(buffer)
                ArrayList<Any?>().apply {
                    repeat(size) { add(readValue(buffer)) }
                }
            }
            MAP -> {
                val size = readSize(buffer)
                HashMap<Any?, Any?>().apply {
                    repeat(size) {
                        put(readValue(buffer), readValue(buffer))
                    }
                }
            }
            else -> throw IllegalArgumentException("Message corrupted: type $type")
        }
    }

    private fun readAlignment(buffer: ByteBuffer, alignment: Int) {
        val mod = buffer.position() % alignment
        if (mod != 0) buffer.position(buffer.position() + alignment - mod)
    }
}

internal class ByteArrayOutputStream {
    private val list = mutableListOf<Byte>()

    fun write(b: Int) { list.add(b.toByte()) }
    fun write(bytes: ByteArray) { list.addAll(bytes.toList()) }
    fun writeShort(v: Int) {
        list.add((v and 0xFF).toByte())
        list.add((v shr 8 and 0xFF).toByte())
    }
    fun writeInt(v: Int) {
        list.add((v and 0xFF).toByte())
        list.add((v shr 8 and 0xFF).toByte())
        list.add((v shr 16 and 0xFF).toByte())
        list.add((v shr 24 and 0xFF).toByte())
    }
    fun writeLong(v: Long) {
        for (i in 0..7) list.add((v shr (i * 8) and 0xFF).toByte())
    }
    fun writeAlignment(alignment: Int) {
        val mod = list.size % alignment
        if (mod != 0) repeat(alignment - mod) { list.add(0) }
    }
    fun toByteArray() = list.toByteArray()
}

File: ios/Classes/StandardMessageCodec.swift
import Foundation

/// Swift implementation of Flutter's StandardMessageCodec binary format
/// (subset: null, bool, int, int64, double, String, List, Map).
enum StandardMessageCodec {
    private static let nullByte: UInt8 = 0
    private static let trueByte: UInt8 = 1
    private static let falseByte: UInt8 = 2
    private static let intByte: UInt8 = 3
    private static let int64Byte: UInt8 = 4
    private static let doubleByte: UInt8 = 6
    private static let stringByte: UInt8 = 7
    private static let listByte: UInt8 = 12
    private static let mapByte: UInt8 = 13

    static func encodeMessage(_ value: Any?) -> Data {
        var stream = DataStream(data: Data())
        writeValue(value, to: &stream)
        return stream.data
    }

    /// Writes a single value to an existing stream (for multi-value messages).
    static func writeValueToStream(_ value: Any?, to stream: inout DataStream) {
        writeValue(value, to: &stream)
    }

    /// Decodes one value from the start of data; returns (value, remainingData).
    static func decodeFirstValue(_ data: Data) -> (value: Any?, remaining: Data) {
        var stream = DataStream(data: data)
        let value = readValue(from: &stream)
        let remaining = stream.position < data.count ? data.suffix(from: stream.position) : Data()
        return (value, Data(remaining))
    }

    private static func writeSize(_ value: Int, to stream: inout DataStream) {
        if value < 254 {
            stream.write(UInt8(value))
        } else if value <= 0xFFFF {
            stream.write(254)
            stream.writeUInt16(UInt16(value))
        } else {
            stream.write(255)
            stream.writeUInt32(UInt32(value))
        }
    }

    private static func writeValue(_ value: Any?, to stream: inout DataStream) {
        switch value {
        case .none:
            stream.write(nullByte)
        case let v as Bool:
            stream.write(v ? trueByte : falseByte)
        case let v as Int:
            if Int32.min <= v && v <= Int32.max {
                stream.write(intByte)
                stream.writeInt32(Int32(v))
            } else {
                stream.write(int64Byte)
                stream.writeInt64(Int64(v))
            }
        case let v as Int64:
            stream.write(int64Byte)
            stream.writeInt64(v)
        case let v as Double:
            stream.write(doubleByte)
            stream.writeAlignment(8)
            stream.writeDouble(v)
        case let v as Float:
            stream.write(doubleByte)
            stream.writeAlignment(8)
            stream.writeDouble(Double(v))
        case let v as String:
            stream.write(stringByte)
            let utf8 = Data(v.utf8)
            writeSize(utf8.count, to: &stream)
            stream.write(utf8)
        case let v as [Any?]:
            stream.write(listByte)
            writeSize(v.count, to: &stream)
            for item in v { writeValue(item, to: &stream) }
        case let v as [AnyHashable: Any?]:
            stream.write(mapByte)
            writeSize(v.count, to: &stream)
            for (k, val) in v {
                writeValue(k.base, to: &stream)
                writeValue(val, to: &stream)
            }
        case let v as AnyHashable:
            writeValue(v.base, to: &stream)
        default:
            if let v = value as? NSNumber {
                stream.write(doubleByte)
                stream.writeAlignment(8)
                stream.writeDouble(v.doubleValue)
            } else {
                fatalError("Unsupported type: \(type(of: value))")
            }
        }
    }

    private static func readSize(from stream: inout DataStream) -> Int {
        let b = Int(stream.readByte())
        switch b {
        case 254: return Int(stream.readUInt16())
        case 255: return Int(stream.readUInt32())
        default: return b
        }
    }

    private static func readValue(from stream: inout DataStream) -> Any? {
        let type = stream.readByte()
        return readValueOfType(type, from: &stream)
    }

    private static func readValueOfType(_ type: UInt8, from stream: inout DataStream) -> Any? {
        switch type {
        case nullByte: return nil
        case trueByte: return true
        case falseByte: return false
        case intByte: return Int(stream.readInt32())
        case int64Byte: return stream.readInt64()
        case doubleByte:
            stream.readAlignment(8)
            return stream.readDouble()
        case stringByte:
            let len = readSize(from: &stream)
            let data = stream.readBytes(len)
            return String(data: data, encoding: .utf8)
        case listByte:
            let count = readSize(from: &stream)
            var list: [Any?] = []
            for _ in 0..<count { list.append(readValue(from: &stream)) }
            return list
        case mapByte:
            let count = readSize(from: &stream)
            var map: [AnyHashable: Any?] = [:]
            for _ in 0..<count {
                let k = readValue(from: &stream)
                let v = readValue(from: &stream)
                if let key = k as? String { map[AnyHashable(key)] = v }
                else if let key = k as? Int { map[AnyHashable(key)] = v }
                else if let key = k as? Bool { map[AnyHashable(key)] = v }
                else if let key = k as? Double { map[AnyHashable(key)] = v }
                else if let key = k as? Int64 { map[AnyHashable(key)] = v }
            }
            return map
        default:
            fatalError("Message corrupted: type \(type)")
        }
    }
}

internal struct DataStream {
    var data: Data
    var position: Int = 0

    init(data: Data) { self.data = data }

    mutating func readByte() -> UInt8 {
        let v = data[position]
        position += 1
        return v
    }

    mutating func readBytes(_ count: Int) -> Data {
        let slice = data.subdata(in: position..<(position + count))
        position += count
        return slice
    }

    mutating func readInt32() -> Int32 {
        let v = data.withUnsafeBytes { $0.load(fromByteOffset: position, as: Int32.self) }
        position += 4
        return Int32(littleEndian: v)
    }

    mutating func readUInt32() -> UInt32 {
        let v = data.withUnsafeBytes { $0.load(fromByteOffset: position, as: UInt32.self) }
        position += 4
        return UInt32(littleEndian: v)
    }

    mutating func readUInt16() -> UInt16 {
        let v = data.withUnsafeBytes { $0.load(fromByteOffset: position, as: UInt16.self) }
        position += 2
        return UInt16(littleEndian: v)
    }

    mutating func readInt64() -> Int64 {
        let v = data.withUnsafeBytes { $0.load(fromByteOffset: position, as: Int64.self) }
        position += 8
        return Int64(littleEndian: v)
    }

    mutating func readDouble() -> Double {
        let u = data.withUnsafeBytes { $0.load(fromByteOffset: position, as: UInt64.self) }
        position += 8
        return Double(bitPattern: UInt64(littleEndian: u))
    }

    mutating func readAlignment(_ alignment: Int) {
        let mod = position % alignment
        if mod != 0 { position += alignment - mod }
    }

    mutating func write(_ byte: UInt8) {
        data.append(byte)
    }

    mutating func write(_ bytes: Data) {
        data.append(bytes)
    }

    mutating func writeInt32(_ v: Int32) {
        var le = v.littleEndian
        data.append(Data(bytes: &le, count: 4))
    }

    mutating func writeUInt32(_ v: UInt32) {
        var le = v.littleEndian
        data.append(Data(bytes: &le, count: 4))
    }

    mutating func writeUInt16(_ v: UInt16) {
        var le = v.littleEndian
        data.append(Data(bytes: &le, count: 2))
    }

    mutating func writeInt64(_ v: Int64) {
        var le = v.littleEndian
        data.append(Data(bytes: &le, count: 8))
    }

    mutating func writeDouble(_ v: Double) {
        var le = v.bitPattern.littleEndian
        data.append(Data(bytes: &le, count: 8))
    }

    mutating func writeAlignment(_ alignment: Int) {
        let mod = data.count % alignment
        if mod != 0 { data.append(Data(count: alignment - mod)) }
    }
}
